aws/resource_aws_kms_grant.go

```diff
@@ -0,0 +1,418 @@
+		}
+	} else {
+		log.Printf("[WARN] %s KMS grant id not found for key id %s, removing from state file", grantId, keyId)
+		d.SetId("")
+	}
+
+	return nil
+}
+
+// Retiring grants requires special permissions (i.e. the
+// caller to be root, retiree principal, or grantee principal with retire grant
+// privileges). So just revoke grants.
+func resourceAwsKmsGrantDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kmsconn
+
+	grantId := d.Get("grant_id").(string)
+	keyId := d.Get("key_id").(string)
+	input := kms.RevokeGrantInput{
+		GrantId: aws.String(grantId),
+		KeyId:   aws.String(keyId),
+	}
+
+	log.Printf("[DEBUG] Revoking KMS grant: %s", grantId)
+	_, err := conn.RevokeGrant(&input)
+
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Checking if grant is revoked: %s", grantId)
+	err = waitForKmsGrantToBeRevoked(conn, keyId, grantId)
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func resourceAwsKmsGrantExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	conn := meta.(*AWSClient).kmsconn
+
+	grantId := d.Id()
+	keyId := d.Get("key_id").(string)
+
+	log.Printf("[DEBUG] Looking for Grant: %s", grantId)
+	grant, err := findKmsGrantByIdWithRetry(conn, keyId, grantId)
+
+	if err != nil {
+		if grant != nil {
+			return true, err
+		}
+		return false, err
+	}
+	if grant != nil {
+		return true, err
+	}
+
+	return false, err
+}
+
+func getKmsGrantById(grants []*kms.GrantListEntry, grantIdentifier string) *kms.GrantListEntry {
+	for idx := range grants {
+		if *grants[idx].GrantId == grantIdentifier {
+			return grants[idx]
+		}
+	}
+
+	return nil
+}
+
+/*
+In the functions below it is not possible to use retryOnAwsCodes function, as there
+is no describe grants call, so an error has to be created if the grant is or isn't returned
+by the list grants call when expected.
+*/
+
+// NB: This function only retries the grant not being returned and some edge cases, while AWS Errors
+// are handled by the findKmsGrantById function
+func findKmsGrantByIdWithRetry(conn *kms.KMS, keyId string, grantId string) (*kms.GrantListEntry, error) {
+	var grant *kms.GrantListEntry
+	err := resource.Retry(2*time.Minute, func() *resource.RetryError {
+		var err error
+		grant, err = findKmsGrantById(conn, keyId, grantId, nil)
+
+		if err != nil {
+			if serr, ok := err.(KmsGrantMissingError); ok {
+				// Force a retry if the grant should exist
+				return resource.RetryableError(serr)
+			}
+
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	return grant, err
+}
+
+// Used by the tests as well
+func waitForKmsGrantToBeRevoked(conn *kms.KMS, keyId string, grantId string) error {
+	err := resource.Retry(2*time.Minute, func() *resource.RetryError {
+		grant, err := findKmsGrantById(conn, keyId, grantId, nil)
+		if err != nil {
+			if _, ok := err.(KmsGrantMissingError); ok {
+				if grant == nil {
```

Comment:
In what condition would the grant be non-nil and the error be non-nil?

Reply:
There isn't one, I'll remove the innermost if statement and just return nil (see findKmsGrantById).